python/ngrok/__main__.py

```diff
@@ -0,0 +1,216 @@
+    tcp_string = None
+    if args.host and args.port:
+        tcp_string = args.host + ":" + str(args.port)
+    elif args.host:
+        tcp_string = args.host + ":" + fallback_port(args)
+    elif args.port:
+        tcp_string = DEFAULT_HOST + ":" + str(args.port)
+    elif (
+        args.bind
+        and not args.bind.startswith("fd://")
+        and not args.bind.startswith("unix:")
+    ):
+        tcp_string = args.bind
+        # check if missing host
+        if re.search("^:\d+$", tcp_string):
+            tcp_string = DEFAULT_HOST + tcp_string
+        # check if missing port
+        if not re.search(":\d+$", tcp_string):
+            tcp_string += ":" + fallback_port(args)
+    return tcp_string
+
+
+def setup_forwarding(tunnel, args):
+    # prefer pipe over tcp
+    pipe_string = get_pipe_string(args)
+    if pipe_string:
+        tunnel.forward_pipe(pipe_string)
+    else:
+        tcp_string = get_tcp_string(args)
+        if not tcp_string:
+            return False
+        tunnel.forward_tcp(tcp_string)
+
+    return True
+
+
+async def bind(parser, args):
+    session = await configure_session(args).connect()
+    tunnel = await configure_tunnel(session, args).listen()
+    tunnel_success = setup_forwarding(tunnel, args)
+
+    # if we don't have what we need, check gunicorn environment variable
+    if (
+        not tunnel_success
+        and args.command == "gunicorn"
+        and "GUNICORN_CMD_ARGS" in os.environ
+    ):
+        env_cmd_args = shlex.split(os.getenv("GUNICORN_CMD_ARGS"))
+        env_args, unknown = parser.parse_known_args(env_cmd_args)
+        tunnel_success = setup_forwarding(tunnel, env_args)
+
+    # fallback to the default host and port for these runners
+    if not tunnel_success:
+        args.host = DEFAULT_HOST
+        args.port = fallback_port(args)
+        tunnel_success = setup_forwarding(tunnel, args)
+
+    # give up
+    if not tunnel_success:
+        logging.fatal("No tunnel created. Exiting.")
+        sys.exit(1)
+
+    return tunnel
+
+
+def main(args):
+    logging.basicConfig(level=logging.INFO)
+
+    # argument parsing
+    parser = ngrok_parser.get_parser()
+    args, unknown = parser.parse_known_args()
+
+    # validation
+    if args.config:
+        logging.fatal("Config file not supported. Exiting.")
+        sys.exit(2)
+    if args.fd or (args.bind and args.bind.startswith("fd://")):
+        logging.fatal("File Descriptor not supported. Exiting.")
+        sys.exit(3)
+
+    # bind to ngrok
+    tunnel = asyncio.run(bind(parser, args))
+
+    # now pretend we don't exist
+    pass_args = [args.command]
+    # pass through some args
+    pass_through_args = ngrok_parser.get_pass_through_args()
+    for key, val in vars(args).items():
+        if key in pass_through_args and val:
+            pass_args.append(f"--{key}")
+            pass_args.append(str(val))
+    # pass through all unknown args
+    sys.argv = pass_args + unknown
+    if args.command == "gunicorn":
```

Comment:
Given that there are a known set of these handled might it be worth checking right after setting the parser and exiting if not handled?

Reply:
Just saw this notification. Yes that covers it!